collect_app/src/main/java/org/odk/collect/android/logic/ImmutableDisplayableQuestion.java

```diff
@@ -0,0 +1,174 @@
+ */
+public class ImmutableDisplayableQuestion {
+    /**
+     * Where in the form this question is located.
+     */
+    private final FormIndex index;
+
+    /**
+     * The main text of the question. This will include a prefixed asterisk if the question is
+     * required.
+     */
+    private final String questionText;
+
+    /**
+     * The hint text of the question.
+     */
+    private final String helpText;
+
+    /**
+     * The guidance hint text of the question.
+     */
+    private final String guidanceText;
+
+    /**
+     * The answer text of the question.
+     */
+    private final String answerText;
+
+    /**
+     * Whether the question is read-only.
+     */
+    private final boolean isReadOnly;
+
+    /**
+     * The choices displayed to a user if this question is of a type that has choices.
+     */
+    private List<SelectChoice> selectChoices;
+
+    /**
+     * Saves all the user-visible aspects of the given {@link FormEntryPrompt}.
+     */
+    public ImmutableDisplayableQuestion(FormEntryPrompt question) {
+        index = question.getIndex();
+        questionText = question.getQuestionText();
+        helpText = question.getHelpText();
+        guidanceText = question.getSpecialFormQuestionText(question.getQuestion().getHelpTextID(), "guidance");
+        answerText = question.getAnswerText();
+        isReadOnly = question.isReadOnly();
+
+        List<SelectChoice> choices = question.getSelectChoices();
+        if (choices != null) {
+            selectChoices = new ArrayList<>();
+            selectChoices.addAll(choices);
+        }
+    }
+
+    public FormIndex getFormIndex() {
+        return index;
+    }
+
+    /**
+     * Returns {@code true} if the provided {@link FormEntryPrompt} has the same user-visible
+     * aspects, {@code false} otherwise.
+     */
+    public boolean sameAs(FormEntryPrompt question) {
+        return question != null
+                && question.getIndex().equals(index)
+                && (question.getQuestionText() == null ? questionText == null : question.getQuestionText().equals(questionText))
+                && (question.getHelpText() == null ? helpText == null : question.getHelpText().equals(helpText))
+                && (getGuidanceHintText(question) == null ? guidanceText == null : getGuidanceHintText(question).equals(guidanceText))
+                && (question.getAnswerText() == null ? answerText == null : question.getAnswerText().equals(answerText))
+                && (question.isReadOnly() == isReadOnly)
+                && selectChoiceListsEqual(question.getSelectChoices(), selectChoices);
+    }
+
+    private static boolean selectChoiceListsEqual(List<SelectChoice> selectChoiceList1, List<SelectChoice> selectChoiceList2) {
+        if (selectChoiceList1 == null) {
+            return selectChoiceList2 == null;
```

Comment:
Isn't it always true?

Reply:
Which part? You could pass in `selectChoiceList1` that is null and `selectChoiceList2` that is not and that should return `false`. This check also acts as a guard for all of the other calls below that dereference `selectChoiceList1`.

Reply:
Ok, my mistake.